pkg/identity/cache/cell/cell.go

```diff
@@ -131,27 +155,96 @@ func newIdentityAllocator(params identityAllocatorParams) identityAllocatorOut {
 // identityAllocatorOwner is used to break the circular dependency between
 // CachingIdentityAllocator and policy.Repository.
 type identityAllocatorOwner struct {
-	policy        policy.PolicyRepository
-	policyUpdater *policy.Updater
+	policy    policy.PolicyRepository
+	epmanager endpointmanager.EndpointManager
 
 	identityHandlers []identity.UpdateIdentities
+
+	// set of notification waitgroups to wait in for batched UpdatePolicyMaps,
+	// and a mutex to protect for writing
+	wgsLock lock.Mutex
+	wgs     []*sync.WaitGroup
+
+	updatePolicyMaps job.Trigger
 }
 
 // UpdateIdentities informs the policy package of all identity changes
-// and also triggers policy updates.
+// and also triggers policy updates. Updates to the endpoints are batched.
 //
 // The caller is responsible for making sure the same identity is not
 // present in both 'added' and 'deleted'.
 func (iao *identityAllocatorOwner) UpdateIdentities(added, deleted identity.IdentityMap) {
+	// Have we already seen this exact set of updates? If so, we can skip.
+	// This happens when a global identity is allocated locally (for an Endpoint).
+	// We will add the identity twice; once directly from the endpoint creation,
+	// and again from the global identity watcher (k8s or kvstore).
+	if iao.policy.GetSelectorCache().CanSkipUpdate(added, deleted) {
+		log.Debug("Skipping no-op identity update")
+		return
+	}
+
+	log.WithFields(logrus.Fields{
+		logfields.AddedPolicyID:   slices.Collect(maps.Keys(added)),
+		logfields.DeletedPolicyID: slices.Collect(maps.Keys(deleted)),
+	}).Info("Processing identity update")
```

Comment:
nit: I suspect we'll want to tune some of these new log messages in high churn environments to avoid flooding the logs too much.

Reply:
So, prior to this change, we would log the following info-level line every time an identity was learned:

```
level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
```

Not particularly useful. Now, we have the same volume of log output, but we also have a changestream, which is invaluable in debugging. I have longed for such a history when diagnosing production policy issues.

Reply:
However, I'd be happy to add a knob to disable this as a follow-up PR.

Reply:
Still on my wishlist is a dynamic logger so we don't have to tune log by log but could leave either up to the user to set subsystems to more or less verbose modes :) but we'll get there.

Other question for me is how long the lists in the fields might be. The key is policyID which sounds small but the value is map keys which sounds potentially very large.

Reply:
The key is just numeric ID, no labels.

However, we could truncate it to avoid spam. Large updates should only come on initial startup, and aren't likely to be interesting regardless.

Reply:
^ related: knowing the count may be as interesting as the individual ids if there are >5